context/ChainsContext/service.tsx

```
@@ -0,0 +1,149 @@
+  return chain;
+};
+
+export const getAssetItemsFromRegistry = async (chainName: string, isTestnet?: boolean) => {
+  const assetsGhPath = isTestnet ? "testnets/" + chainName : chainName;
+  const assetsGhUrl = registryGhUrl + assetsGhPath + "/assetlist.json";
+
+  const response = await fetch(assetsGhUrl);
+  if (!response.ok) {
+    throw new Error(`Failed to get assets for ${chainName} chain from registry`);
+  }
+
+  const assets: readonly RegistryAsset[] = (await response.json()).assets;
+  return assets;
+};
+
+const getNodeFromArray = async (nodeArray: readonly RegistryChainApisRpc[]) => {
+  // only return https connections
+  const secureNodes = nodeArray
+    .filter(({ address }) => address.startsWith("https://"))
+    .map(({ address }) => address);
+
+  if (!secureNodes.length) {
+    throw new Error("No SSL enabled RPC nodes available for this chain");
+  }
+
+  for (const node of secureNodes) {
+    try {
+      // test client connection
+      const client = await StargateClient.connect(node);
+      await client.getHeight();
+      return node;
+    } catch {}
+  }
+
+  throw new Error("No RPC nodes available for this chain");
+};
+
+const getExplorerFromArray = (explorers: readonly RegistryChainExplorer[]) => {
+  return explorers[0]?.tx_page ?? "";
+};
+
+export const getChainFromRegistry = async (chainName: string, isTestnet?: boolean) => {
+  const chainItem = await getChainItemFromRegistry(chainName, isTestnet);
+  const registryAssets = await getAssetItemsFromRegistry(chainName, isTestnet);
+  const firstAsset = registryAssets[0];
+
+  const nodeAddress = await getNodeFromArray(chainItem.apis.rpc);
+  const explorerLink = getExplorerFromArray(chainItem.explorers);
+  const firstAssetDenom = firstAsset.base;
+  const displayDenom = firstAsset.symbol;
+  const displayUnit = firstAsset.denom_units.find((u) => u.denom == firstAsset.display);
+  const displayDenomExponent = displayUnit?.exponent ?? 6; // TODO remove hardcoded fallback 6
```

I think we can just fail if `displayUnit` is not defined. This indicates broken data and requires serious debugging instead of continuing the execution.